File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

for (let line of input) {
  if (line === ".") break;

  const stack = [];

  for (let i = 0; i < line.length; i++) {
    const char = line[i];

    if (char === "(" || char === "[") {
      stack.push(char);
    } else if (char === ")") {
      if (stack.pop() !== "(") {
        console.log("no");
        break;
      }
    } else if (char === "]") {
      if (stack.pop() !== "[") {
        console.log("no");
        break;
      }
    }

    if (i === line.length - 1) {
      if (stack.length === 0) {
        console.log("yes");
      } else {
        console.log("no");
      }
    }
  }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = sys.stdin.readline

while True:
    test = input().rstrip()
    if test == '.':
        break
    
    stack = []

    for char in test:
        if char == '(' or char == '[':
            stack.append(char)
        elif char == ')' or char == ']':
            if stack and ((char == ')' and stack[-1] == '(') or (char == ']' and stack[-1] == '[')):
                stack.pop()
            else:
                print('no')
                break
    else:

        if not stack:
            print('yes')
        else:
            print('no')

       # 업로드 test